Create how-to-use-databricks-workflows-to-run-dbt-cloud-jobs.md (#3102)
fishtown-analytics/docs.getdbt.com: website/docs/guides/dbt-ecosystem/databricks-guides/how-to-use-databricks-workflows-to-run-dbt-cloud-jobs.md

# How to use Databricks workflows to run dbt Cloud Jobs

## Why

> Using Databricks workflows to call the dbt Cloud Job API can be useful for a few reasons:

Reviewer: ```suggestion
Using Databricks workflows to call the dbt Cloud job API can be useful for several reasons:
```

> 1. **Integration with other ETL processes**: If you have other ETL processes running in Databricks, you can use a Databricks workflow to trigger a dbt Cloud job once the ETL processes have completed.

Reviewer: ```suggestion
1. **Integration with other ETL processes**: If you're already running other ETL processes in Databricks, you can use a Databricks workflow to trigger a dbt Cloud job after those processes are done.
```

Reviewer: ```suggestion
1. **Integration with other ETL processes**: If you're already running other ETL processes in Databricks, you can use a Databricks workflow to trigger a dbt Cloud job after those processes are done.
```

> 2. **Utilises dbt Cloud Jobs features:** dbt Cloud gives the ability to monitor jobs, manage historical logs and documentation, optimise model timing and much more.

Reviewer: ```suggestion
2. **Utilizes dbt Cloud jobs features:** dbt Cloud gives the ability to monitor job progress, manage historical logs and documentation, optimize model timing, and much [more](/docs/deploy/dbt-cloud-job).
```

> 3. **Separation of concerns:** detailed logs for dbt jobs will be stored in the dbt Cloud environment which means more modularity and efficient debugging. You will be able to isolate the bugs faster whilst still being able to see the overall status in Databricks.

Reviewer: sugg diff title, does 'detailed logging' encompass what you meant?
```suggestion
3. [**Separation of concerns**](https://en.wikipedia.org/wiki/Separation_of_concerns): Detailed logs for dbt jobs in the dbt Cloud environment can lead to more modularity and efficient debugging. By doing so, it becomes easier to isolate bugs quickly while still being able to see the overall status in Databricks.
```

Author: "Seperation of concerns" is a software engineering phrase: https://en.wikipedia.org/wiki/Separation_of_concerns. Do you still think we should change it or perhaps maybe we should provide a link to it.

Reviewer: oh gotcha, ok cool thank you for clarifying that Mark! no def keep it and I'll sort that for you!

> 4. **Customisable job triggering:** If you want to trigger dbt Cloud jobs based on custom conditions or logic that are not natively supported by dbt Cloud's scheduling feature, you can use a Databricks workflow to trigger the job with a custom script. This gives you more flexibility in how and when your dbt Cloud jobs run.

Reviewer: ```suggestion
4. **Custom job triggering:** Use a Databricks workflow to trigger dbt Cloud jobs based on custom conditions or logic that aren't natively supported by dbt Cloud's scheduling feature. This can give you more flexibility in terms of when and how your dbt Cloud jobs run.
```

> 

Reviewer: @dbt-markwan, do you have to be on a team or enterprise plan since it requires dbt cloud api access?

Author: Yes you'd have to be on Teams or enterprise plan! I've added a pre-req

## Pre-requisites

- Active Databricks Account with access to [Data Science and Engineering workspace](https://docs.databricks.com/workspace-index.html) and [Manage secrets](https://docs.databricks.com/security/secrets/index.html)

> - Existing [dbt Cloud Job](https://docs.getdbt.com/docs/deploy/dbt-cloud-job)

Reviewer: sugg making it clear that it needs to be configured as well
```suggestion
- You must have a configured and existing [dbt Cloud job](/docs/deploy/dbt-cloud-job)
```

- [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html)

> ## How

Reviewer: ```suggestion
## Configure Databricks workflows for dbt Cloud jobs
```

> To use Databricks workflows to run dbt Cloud Jobs, you will need to follow these steps:

Reviewer: ```suggestion
To use Databricks workflows for running dbt Cloud jobs, you need to perform the following steps: 

- [Set up a Databricks secret scope](#set-up-a-databricks-secret-scope)
- [Create a Databricks Python notebook](#create-a-databricks-python-notebook)
- [Configure the workflows to run the dbt Cloud jobs](#configure-workflows)
```

### Part 1: Setup Secrets

> 1. Retrieve [User API Token](https://docs.getdbt.com/docs/dbt-cloud-apis/user-tokens#user-api-tokens) or [Service Account Token](https://docs.getdbt.com/docs/dbt-cloud-apis/service-tokens#generating-service-account-tokens) from dbt Cloud

Reviewer: ```suggestion
1. Retrieve [User API Token](/docs/dbt-cloud-apis/user-tokens#user-api-tokens) or [Service Account Token](/docs/dbt-cloud-apis/service-tokens#generating-service-account-tokens) from dbt Cloud. Copy and save this token.
```

> 2. Set up a Databricks secret scope. This is used to securely store your dbt Cloud API key. You can create a new secret scope by using [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html). Enter the following commands in your terminal:

Reviewer: ```suggestion
2. Set up a Databricks secret scope, which is used to securely store your dbt Cloud API key. You can create a new secret scope by using [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html). 

3. Enter the following commands in your terminal:
```

```bash
# In this example we set up a secret scope and key called "dbt-cloud" and "api-key" respectively.
databricks secrets create-scope --scope <YOUR_SECRET_SCOPE>
databricks secrets put --scope  <YOUR_SECRET_SCOPE> --key  <YOUR_SECRET_KEY> --string-value "<YOUR_DBT_CLOUD_API_KEY>"
```

> Replace **`<YOUR_SECRET_SCOPE>`** and **`<YOUR_SECRET_KEY>`** with your own unique identifiers. Click [here for more information on secrets.](https://docs.databricks.com/security/secrets/index.html)

Reviewer: ```suggestion
4. Replace **`<YOUR_SECRET_SCOPE>`** and **`<YOUR_SECRET_KEY>`** with your own unique identifiers. Click [here](https://docs.databricks.com/security/secrets/index.html) for more information on secrets.
```

> Replace **`<YOUR_DBT_CLOUD_API_KEY>`** with the actual API key value that you copied from dbt Cloud in step 1.

Reviewer: ```suggestion
5. Replace **`<YOUR_DBT_CLOUD_API_KEY>`** with the actual API key value that you copied from dbt Cloud in step 1.
```

### Part 2: Create Databricks Notebook

> 1. [Create a Databricks Python notebook](https://docs.databricks.com/notebooks/notebooks-manage.html). This notebook will execute a Python script that calls the dbt Cloud Job API. 

Reviewer: ```suggestion
1. [Create a Databricks Python notebook](https://docs.databricks.com/notebooks/notebooks-manage.html), which executes a Python script that calls the dbt Cloud job API. 
```

> 2. Create a Python script that calls the dbt Cloud Job API. This script should use the `requests` library to make an `HTTP POST` request to the dbt Cloud Job API endpoint with the appropriate parameters. Here's an example script:

Reviewer: ```suggestion
2. Write a Python script that utilizes the requests library to make an HTTP POST request to the dbt Cloud Job API endpoint using the required parameters. Here's an example script:
```

            # See the API docs for additional parameters that can be passed in,
            # including `schema_override` 
            'cause': f"Triggered by Databricks Workflows.",
        }
    )

    try:
        res.raise_for_status()
    except:
        print(f"API token (last four): ...{api_key[-4:]}")
        raise

    response_payload = res.json()
    return response_payload['data']['id']

def _get_job_run_status(job_run_id):
    res = requests.get(
        url=f"https://{base_url}/api/v2/accounts/{account_id}/runs/{job_run_id}/",
        headers={'Authorization': f"Token {api_key}"},
    )

    res.raise_for_status()
    response_payload = res.json()
    return response_payload['data']['status']

def run():
    job_run_id = _trigger_job()
    print(f"job_run_id = {job_run_id}")   
    while True:
        time.sleep(5)
        status = _get_job_run_status(job_run_id)
        print(DbtJobRunStatus(status))
        if status == DbtJobRunStatus.SUCCESS:
            break
        elif status == DbtJobRunStatus.ERROR or status == DbtJobRunStatus.CANCELLED:
            raise Exception("Failure!")

if __name__ == '__main__':
    run()
```

> Replace **`<YOUR_SECRET_SCOPE>`** and **`<YOUR_SECRET_KEY>`** with the values you used in Part 1.

Reviewer: ```suggestion
3. Replace **`<YOUR_SECRET_SCOPE>`** and **`<YOUR_SECRET_KEY>`** with the values you used [previously](#set-up-a-databricks-secret-scope)
```

> Replace **`<YOUR_BASE_URL>`** and **`<YOUR_ACCOUNT_ID>`** with the correct values of your environment. To find these values, navigate to **dbt Cloud** and select **Deploy > Jobs** and select the Job you want to run. Your URL will be structured like so:

Reviewer: ```suggestion
4. Replace **`<YOUR_BASE_URL>`** and **`<YOUR_ACCOUNT_ID>`** with the correct values of your environment and [Access URL](/docs/cloud/about-cloud/regions-ip-addresses) for your region and plan. To find these values, navigate to **dbt Cloud** and select **Deploy -> Jobs** and select the Job you want to run. Your URL will be structured like so:
```

`https://<YOUR_BASE_URL>/deploy/<YOUR_ACCOUNT_ID>/projects/<YOUR_DBT_PROJECT_ID>/jobs/<YOUR_DBT_CLOUD_JOB_ID>`

> For example:

Reviewer: ```suggestion

```

> `https://cloud.getdbt.com/deploy/000000/projects/111111/jobs/222222`

Reviewer: ```suggestion
For example, if your multi-tenant region is North America, your endpoint is `https://cloud.getdbt.com/deploy/000000/projects/111111/jobs/222222`. 
If your multi-tenant region is APAC, your endpoint is `https://au.dbt.com/deploy/000000/projects/111111/jobs/222222`.
```

Therefore the following can be extracted:

| Parameter | Value  |
|---|---|
| `base_url` | cloud.getdbt.com |
| `account_id` | 000000 |
| `job_id` | 222222 |

> 3. Run the Notebook. This will fail however you should now see a widget at the top of your notebook for `job_id`

Reviewer: ```suggestion
5. Run the Notebook.  It will fail, but you should see a `job_id` widget at the top of your notebook.
```

> 4. In the widget, Enter your`job_id` from step 2.

Reviewer: is step 2 the 'create a python script' step?

```suggestion
6. In the widget, enter your`job_id` from step 2.
```

> 5. Run the Notebook again to kick off the dbt Cloud job. Your results should look similar to the following

Reviewer: ```suggestion
7. Run the Notebook again to trigger the dbt Cloud job. Your results should look similar to the following:
```

```bash
job_run_id = 123456
DbtJobRunStatus.QUEUED
DbtJobRunStatus.QUEUED
DbtJobRunStatus.QUEUED
DbtJobRunStatus.STARTING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.RUNNING
DbtJobRunStatus.SUCCESS
```

> - If needed the job can be cancelled from dbt cloud.

Reviewer: sugg active voice
```suggestion
You can cancel the job from dbt Cloud if necessary.
```

### Part 3: Setup Workflows

> You can set up workflows directly from the notebook OR by adding this notebook to one of your existing workflows

Reviewer: suggest turning this into tabs. if you go with tabs, remember to delete the steps so they're not duplicated.  

```suggestion
You can set up workflows directly from the notebook OR by adding this notebook to one of your existing workflows: 

<Tabs>

<TabItem value="createexisting" label="Create a workflow from existing Notebook">

1. Click **Schedule** on the upper right side of the page
2. Click **Add a schedule**
3. Configure Job name, Schedule, Cluster
4. Add a new parameter called: `job_id` and fill in your Job ID. Refer to [this step](#create-a-databricks-python-notebook) to find your Job ID.
5. Click **Create**
6. To test the Job, Click **Run Now**

</TabItem>

<TabItem value="addexisting" label="Add the Notebook to existing workflow">

1. Open Existing Workflow
2. Click *Tasks*
3. Press “+” icon to add a new task
4. Enter the following:

| Field | Value |
|---|---|
| Task name | `<unique_task_name>` |
| Type | Notebook |
| Source | Workspace |
| Path | `</path/to/notebook>` |
| Cluster | `<your_compute_cluster>` |
| Parameters | `job_id`: `<your_dbt_job_id>` |

5. Select *Save Task*
6. Click *Run Now* to test the workflow

</TabItem>
</Tabs>
```

Author: ooo that's cool!

To create a workflow from an existing Notebook:

1. Click Schedule on the top right
2. Click Add a schedule
3. Configure Job name, Schedule, Cluster
4. Add a new parameter called: `job_id` and fill in your Job ID. Refer to Part 2 Step 2 on how to find your Job ID.
5. Click Create
6. To test the Job, Click Run Now

To add the notebook to an existing Workflow 

1. Open Existing Workflow
2. Click Tasks
3. Press “+” icon to add a new task
4. Enter the following:

| Field | Value |
|---|---|
| Task name | _<unique_task_name>_ |
| Type | Notebook |
| Source | Workspace |
| Path | _</path/to/notebook>_ |
| Cluster | _<your_compute_cluster>_ |
| Parameters | `job_id`: _<your_dbt_job_id>_ |

5. Select Save Task
6. Click Run now to test the workflow

> ## Wrap up

Reviewer: ```suggestion
## Closing
```

> Overall, using Databricks workflows to call the dbt Cloud Job API can help you better integrate your data pipeline processes and schedule more complex workflows.

Reviewer: ```suggestion
Using Databricks workflows to access the dbt Cloud job API can improve integration of your data pipeline processes and enable scheduling of more complex workflows.
```

# How to use Databricks workflows to run dbt Cloud Jobs

Using Databricks workflows to call the dbt Cloud job API can be useful for several reasons:
1. **Integration with other ETL processes**: If you're already running other ETL processes in Databricks, you can use a Databricks workflow to trigger a dbt Cloud job after those processes are done.
2. **Utilizes dbt Cloud jobs features:** dbt Cloud gives the ability to monitor job progress, manage historical logs and documentation, optimize model timing, and much [more](/docs/deploy/dbt-cloud-job).
3. **Separation of concerns:** Detailed logs for dbt jobs in the dbt Cloud environment can lead to more modularity and efficient debugging. By doing so, it becomes easier to isolate bugs quickly while still being able to see the overall status in Databricks.
4. **Custom job triggering:** Use a Databricks workflow to trigger dbt Cloud jobs based on custom conditions or logic that aren't natively supported by dbt Cloud's scheduling feature. This can give you more flexibility in terms of when and how your dbt Cloud jobs run.

## Pre-requisites

> - Active [Teams or Enterprise dbt Cloud Account](https://www.getdbt.com/pricing/)

Reviewer: ```suggestion
- Active [Teams or Enterprise dbt Cloud account](https://www.getdbt.com/pricing/)
```

- You must have a configured and existing [dbt Cloud job](/docs/deploy/dbt-cloud-job)
- - Active Databricks Account with access to [Data Science and Engineering workspace](https://docs.databricks.com/workspace-index.html) and [Manage secrets](https://docs.databricks.com/security/secrets/index.html)
- [Databricks CLI](https://docs.databricks.com/dev-tools/cli/index.html). 
  - Please note that you only need to set up your authentication. After you have setup Host and Token and are able to run `databricks workspace ls /Users/<someone@example.com>` you will be able to complete the rest of this guide.

## Configure Databricks workflows for dbt Cloud jobs

> ## Configure Databricks workflows for dbt Cloud jobs

Reviewer: ```suggestion
```

# How to use Databricks workflows to run dbt Cloud Jobs

> ## Why

Reviewer: ```suggestion
```
